ui/index.d.ts

```diff
@@ -171,20 +171,26 @@ export declare abstract class BaseEditor<ContextT> implements Disposable {
 export declare class AzureActionHandler {
     constructor(extensionContext: ExtensionContext, outputChannel: OutputChannel, telemetryReporter?: TelemetryReporter);
 
-    registerCommand(commandId: string, callback: (...args: any[]) => any): void
-    registerCommandWithCustomTelemetry(commandId: string, callback: (properties: TelemetryProperties, measurements: TelemetryMeasurements, ...args: any[]) => any): void;
+    registerCommand(commandId: string, callback: (this: IActionContext, ...args: any[]) => any): void;
 
     /**
-     * NOTE: An event callback must call trackTelemetry() in order for it to be "opted-in" for telemetry. This is meant to be called _after_ the event has determined that it apples to that extension
-     * For example, we might not want to track _every_ onDidSaveEvent, just the save events for our files
+     * NOTE: By default, this sends a telemetry event every single time the event fires. It it recommended to use 'this.sendTelemetry' to only send events if they apply to your extension
      */
-    registerEvent<T>(eventId: string, event: Event<T>, callback: (trackTelemetry: () => void, ...args: any[]) => any): void;
-    registerEventWithCustomTelemetry<T>(eventId: string, event: Event<T>, callback: (trackTelemetry: () => void, properties: TelemetryProperties, measurements: TelemetryMeasurements, ...args: any[]) => any): void;
+    registerEvent<T>(eventId: string, event: Event<T>, callback: (this: IActionContext, ...args: any[]) => any): void;
 
     /**
-     * Use this method for generic calls that you want to track telemetry for, but that aren't registered with VS Code
+     * Use these methods for generic calls that you want to handle, but that aren't registered with VS Code
      */
-    callWithTelemetry(callbackId: string, callback: (properties: TelemetryProperties, measurements: TelemetryMeasurements) => any): Promise<any>;
+    static callWithTelemetry(callbackId: string, telemetryReporter: TelemetryReporter | undefined, callback: (this: IActionContext) => any): Promise<any>;
```

Comment:
I still think these should be extracted from AzureActionHandler, they seem completely unrelated. I had it as a separate top-level function. Maybe a namespace?

Reply:
Thinking about this...  I like how it gets rid of callWithCustomTelemetry

Reply:
Not sure we're going to agree on this one. They share 90% of the same code and this 'handles azure actions'

Reply:
I also don't think telemetry is just about actions. You could have telemetry about performance metrics that are unrelated to any specific action, for instance.

But it's not a big enough deal to worry about. Still an improvement.

Reply:
If "Action" is the part you're worried about, then I'm happy to rename the class - but I do prefer to have the code together

Reply:
I wasn't feeling the "completely unrelated" argument, but you won me over with the "I don't like static methods" argument. Moved to separate file 😊

Reply:
😆